File: timer.py
import time

class Timer:
    def __init__(self):
        self._duration_ms = 0
        self._start_time_monotonic = 0
        self._stop_time_monotonic = 0
        self._elapsed_ms = 0
        self._is_running = False
        self._is_idle = True

    def set_duration(self, minutes, seconds, milliseconds=0):
        self._duration_ms = (minutes * 60000) + (seconds * 1000) + milliseconds

    def start(self):
        if not self._is_running:
            self._is_idle = False
            now = time.monotonic() * 1000  # Convert to ms
            
            if self._elapsed_ms > 0:
                self._start_time_monotonic = now - self._elapsed_ms
            else:
                self._start_time_monotonic = now
            self._is_running = True

    def stop(self):
        if self._is_running:
            self._stop_time_monotonic = time.monotonic() * 1000
            self._elapsed_ms = self._stop_time_monotonic - self._start_time_monotonic
            self._is_running = False

    def reset(self):
        self._is_running = False
        self._is_idle = True
        self._start_time_monotonic = 0
        self._stop_time_monotonic = 0
        self._elapsed_ms = 0

    @property
    def is_running(self):
        return self._is_running

    @property
    def is_paused(self):
        return not self._is_running and not self._is_idle

    @property
    def is_idle(self):
        return self._is_idle

    def is_expired(self):
        return self.get_current_elapsed_ms() >= self._duration_ms

    def get_current_elapsed_ms(self):
        if self._is_running:
            return (time.monotonic() * 1000) - self._start_time_monotonic
        return self._elapsed_ms

    def get_remaining_ms(self):
        elapsed = self.get_current_elapsed_ms()
        if elapsed >= self._duration_ms:
            return 0
        return self._duration_ms - elapsed

    def get_duration_ms(self):
        return self._duration_ms

    @staticmethod
    def ms_to_components(ms):
        minutes = int(ms // 60000)
        ms %= 60000
        seconds = int(ms // 1000)
        ms %= 1000
        return {
            "minutes": minutes,
            "seconds": seconds,
            "milliseconds": int(ms)
        }

File: src/Timer.cpp
/**
 * Source code for timer calculations and management
 */

#include "Timer.h"
#include <Arduino.h>  // For millis()

Timer::Timer()
    : _duration_ms(0),
      _start_time_ms(0),
      _stop_time_ms(0),
      _elapsed_ms(0),
      _is_running(false),
      _is_idle(true)
{
}

void Timer::setDuration(Components duration)
{
    _duration_ms = componentsToMilliseconds(duration);
}

void Timer::start()
{
    if (!_is_running)
    {
        _is_idle = false;  // No longer idle once started
        
        // If resuming from a stop/pause, adjust start time
        if (_elapsed_ms > 0)
        {
            _start_time_ms = millis() - _elapsed_ms;
        }
        else
        {
            _start_time_ms = millis();
        }
        _is_running = true;
    }
}

void Timer::stop()
{
    if (_is_running)
    {
        _stop_time_ms = millis();
        _elapsed_ms = _stop_time_ms - _start_time_ms;
        _is_running = false;
    }
}

void Timer::reset()
{
    _is_running = false;
    _is_idle = true;  // Back to idle state
    _start_time_ms = 0;
    _stop_time_ms = 0;
    _elapsed_ms = 0;
}

Timer::Components Timer::getElapsedTime()
{
    unsigned long current_elapsed;
    
    if (_is_running)
    {
        current_elapsed = millis() - _start_time_ms;
    }
    else
    {
        current_elapsed = _elapsed_ms;
    }
    
    return millisecondsToComponents(current_elapsed);
}

Timer::Components Timer::getRemainingTime()
{
    unsigned long current_elapsed;
    
    if (_is_running)
    {
        current_elapsed = millis() - _start_time_ms;
    }
    else
    {
        current_elapsed = _elapsed_ms;
    }
    
    // If elapsed exceeds duration, remaining is 0
    if (current_elapsed >= _duration_ms)
    {
        return {0, 0, 0};
    }
    
    unsigned long remaining = _duration_ms - current_elapsed;
    return millisecondsToComponents(remaining);
}

Timer::Components Timer::getDuration()
{
    return millisecondsToComponents(_duration_ms);
}

bool Timer::isRunning()
{
    return _is_running;
}

bool Timer::isPaused()
{
    return !_is_running && !_is_idle;
}

bool Timer::isIdle()
{
    return _is_idle;
}

bool Timer::isExpired()
{
    unsigned long current_elapsed;
    
    if (_is_running)
    {
        current_elapsed = millis() - _start_time_ms;
    }
    else
    {
        current_elapsed = _elapsed_ms;
    }
    
    return current_elapsed >= _duration_ms;
}

Timer::Components Timer::millisecondsToComponents(unsigned long ms)
{
    Components result;
    
    result.minutes = ms / 60000;
    ms %= 60000;
    
    result.seconds = ms / 1000;
    ms %= 1000;
    
    result.milliseconds = ms;
    
    return result;
}

unsigned long Timer::componentsToMilliseconds(const Components &components)
{
    unsigned long total = 0;
    
    total += components.minutes * 60000UL;
    total += components.seconds * 1000UL;
    total += components.milliseconds;
    
    return total;
}